Add capacity to override MON ports

pgoron opened the pull request:
Required when you want to host MONs from different ceph clusters
on same nodes.

Commit:
Add capacity to override MON ports

Required when you want to host MONs from different ceph clusters
on same nodes.

pgoron commented on pkg/operator/ceph/config/store.go:
```diff
@@ -61,22 +61,22 @@ func GetStore(context *clusterd.Context, namespace string, ownerInfo *k8sutil.Ow
 
 // CreateOrUpdate creates or updates the stored Ceph config based on the cluster info.
-func (s *Store) CreateOrUpdate(clusterInfo *cephclient.ClusterInfo) error {
+func (s *Store) CreateOrUpdate(clusterInfo *cephclient.ClusterInfo, msgr2Port int32) error {
```
Not really satisfied by the approach for handling msgr2Port override.  Ideally it should be part of clusterInfo type but it triggers too much changes.

geobeau replied:
Maybe we can just change MonInfo.
```
// NewMonInfo returns a new Ceph mon info struct from the given inputs.
func NewMonInfo(name, ip string, port int32) *MonInfo {
	return &MonInfo{Name: name, Endpoint: net.JoinHostPort(ip, fmt.Sprintf("%d", port))}
}
```
to
```
// NewMonInfo returns a new Ceph mon info struct from the given inputs.
func NewMonInfo(name, ip string, port int32, portv2 int32) *MonInfo {
	return &MonInfo{Name: name, Endpoint: net.JoinHostPort(ip, fmt.Sprintf("%d", port), portv2)}
}
```

The NewMonInfo is called in method with access to the spec so it might be possible to inject it here

pgoron replied:
NewMonInfo isn't the only place where MonInfo are created (see ParseMonEndpoints where cluster spec is not available).

geobeau commented:
If you look in the types.go file for `kubebuilder:default=` it seems that there is support for default value, it would help removing the if and have the spec as source of thruth

Commit:
Add mon default ports in CRD

pgoron commented:
> If you look in the types.go file for `kubebuilder:default=` it seems that there is support for default value, it would help removing the if and have the spec as source of truth

Done

Commit:
Retrieve msgr2Port from cluster spec

Commit:
Use mon tcp-msgr2 port value from cluster spec

Commit:
Use mon ports defined in cluster spec

Commit:
Enforce usage of public-addrv to specify bind points of MONs

When not specifying protocol version via vector form, ceph mons will
automatically detect protocol version to use based on default port. They
will consider anyport different from 6789 to be v2 by default.

Commit:
Enforce passage of MonSpec at ClusterInfo creation

geobeau commented:
Looks good
